File: frontend/src/app/user_db.services/user-data.service.ts
import {Injectable} from '@angular/core';
import {SERVER_ADDRESS} from '../globals';
import {HttpClient, HttpHeaders} from '@angular/common/http';
import {UtilitiesService} from '../services/utilities.service';
import {map, Observable} from 'rxjs';
import {UserDataDto} from '../user_db.dto/user-data.dto';
import {UUID} from 'node:crypto';

@Injectable({providedIn: 'root'})
export class UserDataService {

  private DOMAIN = SERVER_ADDRESS + 'userData';

  constructor(private httpClient: HttpClient,
              private utilitiesService: UtilitiesService) {
  }

  public getAllUserData(): Observable<UserDataDto[]> {
    return this.httpClient.get<any[]>(this.DOMAIN)
      .pipe(
        map(data => data.map(item => {
          return new UserDataDto(item.user_data_id, item.user_id, item.role_name, item.status, item.cre_dat);
        }))
      );
  }

  public getUserDataByUserId(user_id: UUID): Observable<UserDataDto> {
    return this.httpClient.get<any>(this.DOMAIN + '/user_id=' + user_id)
      .pipe(
        map(item => {
          return new UserDataDto(item.user_data_id, item.user_id, item.role_name, item.status, item.cre_dat);
        })
      );
  }

  public saveUserData(userData: UserDataDto): Observable<UserDataDto> {

    const userDataToSend = {
      user_data_id: userData.getUserDataId(),
      user_id: userData.getUserId(),
      role_name: userData.getRoleName(),
      status: userData.getStatus(),
      cre_dat: this.utilitiesService.formatDate(userData.getCreDat())
    };

    return this.httpClient.post<UserDataDto>(
      this.DOMAIN + '/save',
      userDataToSend,
      {
        headers: new HttpHeaders({
          'Content-Type': 'application/json'
        })
      }
    );
  }

}

File: user-db/src/main/java/group_05/ase/user_db/services/UserDataService.java
package group_05.ase.user_db.services;

import group_05.ase.user_db.entities.UserDataEntity;
import group_05.ase.user_db.repositories.UserDataRepository;
import group_05.ase.user_db.restData.UserDataDTO;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

@Service
public class UserDataService {

    private final UserDataRepository repository;

    public UserDataService(UserDataRepository repository) {
        this.repository = repository;
    }

    public List<UserDataDTO> getAllUserData() {

        ArrayList<UserDataDTO> userData = new ArrayList<>();
        List<UserDataEntity> tmp = this.repository.findAll();

        for(UserDataEntity data : tmp) {
            userData.add(new UserDataDTO(data.getUserDataId(), data.getUserId(), data.getRoleName(),
                    data.getStatus(), data.getCreDat()));
        }

        return userData;

    }

    public UserDataDTO getUserDataById(int userDataId) {

        UserDataEntity tmp = this.repository.findByUserDataId(userDataId);

        return new UserDataDTO(tmp.getUserDataId(), tmp.getUserId(), tmp.getRoleName(), tmp.getStatus(), tmp.getCreDat());

    }

    public UserDataDTO getUserDataByUserId(UUID userId) {

        UserDataEntity tmp = this.repository.findByUserId(userId);

        return new UserDataDTO(tmp.getUserDataId(), tmp.getUserId(), tmp.getRoleName(), tmp.getStatus(), tmp.getCreDat());

    }

    public UserDataDTO saveUserData(UserDataDTO userDataDTO) {

        UserDataEntity tmp = this.repository.findByUserId(userDataDTO.getUserId());

        if(tmp == null) {
            tmp = new UserDataEntity();

            tmp.setUserId(userDataDTO.getUserId());
            tmp.setRoleName(userDataDTO.getRoleName());
            tmp.setStatus(userDataDTO.getStatus());
            tmp.setCreDat(userDataDTO.getCreDat());

            tmp = this.repository.save(tmp);
        } else {
            tmp.setRoleName(userDataDTO.getRoleName());
            tmp.setStatus(userDataDTO.getStatus());

            tmp = this.repository.save(tmp);
        }

        return new UserDataDTO(tmp.getUserDataId(), tmp.getUserId(), tmp.getRoleName(), tmp.getStatus(), tmp.getCreDat());

    }

}
